### PR TITLE
feat: add IO states to state representation (cpp)

### PR DESCRIPTION
**Pull request by domire8**
## Description

- #156 

This turns out to be quite the PR, but its honestly a lot of boilderplate code and docstrings. The implementation of the classes follow our common patterns and don't add much complexity or new features. Same for the tests, it's the same lines as in the other state classes, to test all setters and getters.

Protocol and pybindings come in follow up PRs

## Review guidelines

Estimated Time of Review: 15 minutes

## Checklist before merging:

- [x] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

### CODE REVIEW

File: source/state_representation/include/state_representation/IOState.hpp [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  void set_names(unsigned int nb_ios);
+
+  /**
+   * @brief Setter of the names from a list of IO names
+   * @param names The vector of strings containing the IO names
+   */
+  void set_names(const std::vector<std::string>& names);
+
+  /**
+   * @brief Set the vlues of the IO state from a single Eigen vector
```
Comment (eeberhard) [inline source/state_representation/include/state_representation/IOState.hpp:56]:
```suggestion
   * @brief Set the values of the IO state from a single Eigen vector
```

File: source/state_representation/include/state_representation/IOState.hpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  friend void swap(IOState& state1, IOState& state2) {
+    swap(static_cast<State&>(state1), static_cast<State&>(state2));
+    std::swap(state1.names_, state2.names_);
+    std::swap(state1.data_, state2.data_);
+  }
+
+  static void assert_index_in_range(unsigned int io_index, unsigned int size);
+
+  std::vector<std::string> names_;///< names of the IOs
+  Eigen::Vector<T, Eigen::Dynamic> data_;///< IO values
```
Comment (eeberhard) [inline source/state_representation/include/state_representation/IOState.hpp:113]: Any particular reason you are using an Eigen vector as the data container type? I guess for conventions across the other class and enabling the `data()` and `to_std_vector()` methods?

File: source/state_representation/include/state_representation/AnalogIOState.hpp
Comment (eeberhard) [inline source/state_representation/include/state_representation/AnalogIOState.hpp:1]: Not sure how you managed, but you mixed up analog and digital 😅 digital means discrete logic, 1s and 0s, while analog means continuous values.

At first I was going to leave a single comment like this that you got the template type wrong, 

```diff
--class AnalogIOState : public IOState<bool>
++class AnalogIOState : public IOState<double>
```

but actually looking at the methods you implemented `is_false` etc here, so really you need to rename the whole class, file and docstrings.

File: source/state_representation/include/state_representation/StateType.hpp [hunk 3]
```diff
@@ -29,6 +29,8 @@ enum class StateType {
   GEOMETRY_SHAPE,
   GEOMETRY_ELLIPSOID,
   TRAJECTORY,
+  ANALOG_IO_STATE,
+  DIGITAL_IO_STATE,
```
Comment (eeberhard) [inline source/state_representation/include/state_representation/StateType.hpp:33]: the most minor of nitpicks but I would prefer `digital` to come before `analog`. It's a little like the elevation of `bool` to `double` in the order of parameter types, but also more common to how you would say "digital and analog IOs"

**Review by eeberhard — COMMENTED** [review]
Looks pretty good! A lot of boilerplate indeed and plenty of functions that might be a bit redundant, but it does fit in very nicely with other state types. 

Just a shame that you mixed up the names analog and digital, but nothing some smart renaming won't fix!

**Comment by domire8** [comment]
Hahaha that is because I really thought analog is 1/0 and digital anything😁 well.....

File: source/state_representation/include/state_representation/IOState.hpp (hunk 2 above)
Comment (domire8) [reply source/state_representation/include/state_representation/IOState.hpp:113]: Yes indeed, for that reason exactly.

**Review by bpapaspyros — DISMISSED** [review]

**Review by bpapaspyros — APPROVED** [review]

File: source/state_representation/include/state_representation/StateType.hpp (hunk 3 above)
Comment (eeberhard) [reply source/state_representation/include/state_representation/StateType.hpp:33]: you marked this as resolved but if it doesn't cause any other issues I would still prefer it the other way around

File: source/state_representation/src/AnalogIOState.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+AnalogIOState::AnalogIOState(const AnalogIOState& state) : AnalogIOState(state.get_name(), state.get_names()) {
+  if (state) {
+    this->set_data(state.data());
+  }
+}
+
+AnalogIOState AnalogIOState::Zero(const std::string& name, unsigned int nb_ios) {
+  AnalogIOState zero = AnalogIOState(name, nb_ios);
+  // as opposed to the constructor specify this state to be filled
```
Comment (eeberhard) [inline source/state_representation/src/AnalogIOState.cpp:31]: nitpick: the phrasing of the comment is a bit weird
```suggestion
  // specify that the default constructed zero state is non-empty
```

File: source/state_representation/src/DigitalIOState.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  DigitalIOState zero = DigitalIOState(name, io_names);
+  // as opposed to the constructor specify this state to be filled
+  zero.set_empty(false);
+  return zero;
+}
+
+DigitalIOState DigitalIOState::Random(const std::string& name, unsigned int nb_ios) {
+  DigitalIOState random = DigitalIOState(name, nb_ios);
+  // set all the state variables to random
+  random.set_data(Eigen::Vector<bool, -1>::Random(random.get_size()));
```
Comment (eeberhard) [inline source/state_representation/src/DigitalIOState.cpp:46]: I assume the `-1`  is the template value that specifies dynamic sizing. I would prefer you use the special "keyword" `Eigen::Dynamic` for readability and to conform with their intended API

**Review by eeberhard — COMMENTED** [review]
Just a few nitpicks, and needs to be rebased with newer version, but I think this will be good to go!

File: source/state_representation/include/state_representation/StateType.hpp (hunk 3 above)
Comment (domire8) [reply source/state_representation/include/state_representation/StateType.hpp:33]: Oups sorry, I think I did on the proto branch instead.

File: source/state_representation/src/AnalogIOState.cpp (hunk 4 above)
Comment (domire8) [reply source/state_representation/src/AnalogIOState.cpp:31]: This is legacy phrasing :smile:

**Review by eeberhard — APPROVED** [review]
Very cool

**Pull request merged**